File: 207-course-schedule/207-course-schedule.cpp
class Solution {
public:
    bool hasCycle(
        int node, vector<vector<int>>& adj, bool visited[],
        bool dfs[]
    ){
        dfs[node]=1;
        visited[node]=1;
        for(auto neighbor: adj[node]){
            if(!visited[neighbor]){
              if(hasCycle(neighbor,adj,visited,dfs)){
                  return true;
              }   
            }
            else{
               if(dfs[neighbor]) return true; 
            }
        }
        
        dfs[node]=0;
        return false;
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        bool visited[numCourses], dfs[numCourses];
        memset(visited,false,numCourses);
        memset(dfs,false,numCourses);
        
        for(auto p:prerequisites){
            adj[p[1]].push_back(p[0]);
        }
        
        
        for(int i=0;i<numCourses;i++){
            if(!visited[i]){
                if(hasCycle(i,adj,visited,dfs)){
                    return false;
                }
            }
        }
        
        return true;
    }
};

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, V: int, pr: List[List[int]]) -> bool:
        adj = defaultdict(set)
        for a,b in pr:
            adj[b].add(a)
        
        vis = {}
        def dfs(node):
            if node in vis:
                return vis[node] # true if cycle is there
            vis[node]=True
            for dest in adj[node]:
                if dfs(dest):
                    return True
            vis[node]=False
            return False
        for i in range(V):
            if dfs(i):
                return False
        return True
        
#         q,ans,indegree = deque(),[],{i:0 for i in range(V)}
#         for i in range(V):
#             for j in adj[i]: indegree[j]+=1
#         for i in range(V):
#             if indegree[i]==0: q.append(i)
#         cnt=0
#         while q:
#             cur = q.popleft()
#             ans.append(cur)
#             for dest in adj[cur]:
#                 indegree[dest]-=1
#                 if indegree[dest]==0: q.append(dest)
#             cnt+=1
#         return cnt!=V:
        
        